File: nve.worker.cpp
#include <fstream>
#include <iostream>
#include <vector>

bool endRunning();
void next();
void runall();

uint32_t size = 0;
uint8_t* pdata;

uint32_t memr_length = 100;
uint32_t* memr;
uint32_t progcnt = 0;
uint32_t stackp = 0;
uint32_t framp = 0;

int main(int argc,char* argv[]) {
    char* fname = argv[1];
    std::cout << "File: " << fname << std::endl;
    std::ifstream ifs(fname, std::ios::in|std::ios::binary);
    if (ifs.fail()) {
        std::cerr << "Failed to open file." << std::endl;
        return 1;
    }
    uint8_t _data;
    int i=0;
    char* mn = (char*)"nveof\n";
    while (!ifs.eof()&&i<10) {
        ifs.read((char*)&_data,sizeof(uint8_t));
        if (i<6&&_data!=mn[i]) {
            std::cout << "This file is not NVE-OF" << std::endl;
            return 1;
        }
        else if (i>5) {
            size += _data*(1<<(i-6)*8);
        }
        i++;
    }
    std::cout << "size: " << size << std::endl;
    pdata = (uint8_t*)malloc(size*5);
    i=0;
    while (!ifs.eof()&&i<size*5) {
        ifs.read((char*)&_data,sizeof(uint8_t));
        pdata[i] = _data;
        i++;
    }
    // for (int a=0;a<size*5;a++) {
    //     std::cout << (int)pdata[a] << " ";
    // }
    memr = (uint32_t*)malloc(sizeof(uint32_t)*memr_length);
    runall();
    return 0;
}

int prog(uint32_t i) {
    return (int)pdata[i*5];
}
uint32_t imme(uint32_t i) {
    return (int)pdata[i*5+1]+(int)pdata[i*5+2]*0x100+(int)pdata[i*5+3]*0x10000+(int)pdata[i*5+4]*0x1000000;
}

void push(uint32_t i) {
    memr[stackp] = i;
    stackp++;
}
uint32_t pop() {
    memr[stackp] = 0;
    stackp--;
    return memr[stackp];
}

void runall() {
    int cnt = 0;
    while (cnt<100000&&!endRunning()) {
        cnt++;
        next();
    }
}
bool endRunning() {
    return progcnt>=size;
}

void next() {
    if (endRunning()) {
        std::cerr << "end runnning" << std::endl;
    }

    //std::cout << prog(progcnt) << " " << imme(progcnt) << " " << progcnt << std::endl;

    uint64_t cs;
    switch (prog(progcnt))
    {
        case 0: // push
            push(imme(progcnt));
        break;
        case 1: // pop n
            for (int i=0;i<imme(progcnt);i++) {
                pop();
            }
        break;
        case 2: // call
            push(framp);
            framp = stackp;
            push(progcnt);
            progcnt = imme(progcnt);
        break;
        case 3: // ret
            progcnt = pop();
            framp = pop();
        break;
        case 4: // fram m
            for (int i=0;i<imme(progcnt);i++) {
                push(0);
            }
        break;
        case 5: // setvar a
            memr[(framp+imme(progcnt))&(4294967295)] = pop();
        break;
        case 6: // getvar a
            push(memr[(framp+imme(progcnt))&(4294967295)]);
        break;
        case 7: // setgvar a
            memr[(imme(progcnt))&(4294967295)] = pop();
        break;
        case 8: // getgvar
            push(memr[(imme(progcnt))&(4294967295)]);
        break;
        case 9: // setdata a
            memr[memr_length-(imme(progcnt)&4294967295)] = pop();
        break;
        case 10: // getdata a
            push(memr[memr_length-(imme(progcnt)&4294967295)]);
        break;
        case 11: // jmp
            progcnt = imme(progcnt);
        break;
        case 12: // ifjmp
            if (pop!=0) {
                progcnt = imme(progcnt);
            }
        break;
        case 13: // add
            push(pop()+pop());
        break;
        case 14: // addc
            cs = pop()+pop()+pop();
            push(cs>>16>>16);
            push(cs&(4294967295));
        break;
        case 15: // and
            push(pop()&pop());
        break;
        case 16: // or
            push(pop()|pop());
        break;
        case 17: // xor
            push(pop()^pop());
        break;
        case 18: // equal
            push(pop()==pop());
        break;
        case 19: // less
            push(pop()<pop());
        break;
        case 20: // greater
            push(pop()>pop());
        break;
        case 21: // enot
            push(~pop());
        break;
        case 22: // notb
            push(!pop());
        break;
        case 23: // out
            std::cout << "[output] " << std::hex << pop() << std::endl;
        break;
        default:
        break;
    }
    progcnt++;
}

File: nve.worker.js
const fs = require("fs");
class NVE {
    #memr;#progcnt;#stackp;#framp; // 宣言
    constructor(file,args=[0]) { // 初期化
        this.file = file;
        this.check(file);
        let _size = new Uint8Array((this.file.slice(6,10)))
        this.size = _size[0]+_size[1]*0x100+_size[2]*0x10000+_size[3]*0x1000000;
        console.log("Size:",this.size.toString(10))
        this.#memr = new Uint32Array(2**32); // 実行用スタック
        this.#progcnt = 0;
        this.#stackp = 1;
        this.#framp = 0;
    }
    prog(i) {
        let a = new Uint8Array(this.file.slice(10+i*5,10+i*5+1))
        return a[0];
    }
    imme(i) {
        let a = new Uint8Array(this.file.slice(10+i*5+1,10+i*5+5))
        return a[0]+a[1]*0x100+a[2]*0x10000+a[3]*0x1000000;
    }
    push(x) {
        this.#memr[this.#stackp] = x;
        this.#stackp++;
    }
    pop() {
        this.#memr[this.#stackp] = 0;
        this.#stackp--;
        return this.#memr[this.#stackp];
    }
    next() { // 一つの命令を実行する
        if (this.endRunning()) {console.warn("end runnning")}
        //console.log(this.#progcnt,this.#prog[this.#progcnt],this.#imme[this.#progcnt])
        // let memshow = function(mem,sp,fp) {
        //     let ret = "";
        //     for (let i=0;i<mem.length;i++) {
        //         if (fp==i) {ret += ","}else {ret += " "}
        //         if (sp==i) {ret += "."}else {ret += " "}
        //         let memi = mem[i];
        //         let n = memi.toString(16);
        //         ret += "00000000".slice(n.length);
        //         ret += n;
        //     }
        //     return ret
        // }
        // let ins = ["push","pop","call","ret","fram","setvar","getvar","setgvar","getgvar","setdata","getdata","jmp","ifjmp","add","addc","and","or","xor","equal","less","greater","not","notb","out"];
        // console.log("[internal state]"," mem:",memshow(this.#memr.slice(0,17),this.#stackp,this.#framp)," pc:",this.#progcnt.toString(16)," sp:",this.#stackp.toString(16)," fp:",this.#framp.toString(16))
        // console.log("")
        // console.log("[next]"," opcode:",this.prog(this.#progcnt).toString(16)," mnemonic:",ins[this.prog(this.#progcnt)]," immediate:",this.imme(this.#progcnt).toString(16))
        // let adr = function(n) {return n&((2**32)-1);};
        switch (this.prog(this.#progcnt)) {
            case 0: // push n スタックに即値を入れる
                this.push(this.imme(this.#progcnt));
            break;
            case 1: // pop n スタックトップの値をn個消す
                for (let i=0;i<this.imme(this.#progcnt);i++) {
                    this.pop();
                }
            break;
            case 2: // call 関数を呼ばれたとき
                this.push(this.#framp);
                this.#framp = this.#stackp;
                this.push(this.#progcnt);
                this.#progcnt = this.imme(this.#progcnt);
            break;
            case 3: // ret 関数から返るとき
                this.#progcnt = this.pop();
                this.#framp = this.pop();
                for (let i=0;i<this.imme(this.#progcnt);i++) {
                    this.pop();
                }
            break;
            case 4: // fram m 局所変数の領域を確保する 0埋めする
                for (let i=0;i<this.imme(this.#progcnt);i++) {
                    this.push(0);
                }
            break;
            // データ移動命令
            case 5: // setvar a a個目の局所変数に値を入れる
                this.#memr[(this.#framp+this.imme(this.#progcnt))&(4294967295)] = this.pop();
            break;
            case 6: // getvar a a個目の局所変数から値を複製する
                this.push(this.#memr[(this.#framp+this.imme(this.#progcnt))&(4294967295)]);
            break;
            case 7: // setgvar a a個目のグローバル変数に値を入れる
                this.#memr[(this.imme(this.#progcnt))&(4294967295)] = this.pop();
            break;
            case 8: // getgvar a a個目のグローバル変数から値を複製する
                this.push(this.#memr[(this.imme(this.#progcnt))&(4294967295)]);
            break;
            case 9: // setdata a a個目のヒープ領域に値を入れる
                this.#memr[this.#memr.length-((this.imme(this.#progcnt))&(4294967295))] = this.pop();
            break;
            case 10: // getdata a a個目のヒープ領域から値を複製する
                this.push(this.#memr[this.#memr.length-((this.imme(this.#progcnt))&(4294967295))]);
            break;
            // ジャンプ命令
            case 11: // jmp
                this.#progcnt = this.imme(this.#progcnt);
            break;
            case 12: // ifjmp
                if (this.pop()!=0) {
                    this.#progcnt = this.imme(this.#progcnt);
                }
            break;

            // 2項算術演算
            case 13: // add
                this.push(this.pop()+this.pop());
            break;
            case 14: // addc
                let cs = this.pop()+this.pop()+this.pop();
                this.push(cs>>>32);
                this.push(cs&(4294967295));
            break;
            // 2項ビット演算
            case 15: // and
                this.push(this.pop()&this.pop());
            break;
            case 16: // or
                this.push(this.pop()|this.pop());
            break;
            case 17: // xor
                this.push(this.pop()^this.pop());
            break;
            // 2項比較演算
            case 18: // equal
                this.push(Number(this.pop()==this.pop()));
            break;
            case 19: // less
                this.push(Number(this.pop()>this.pop()));
            break;
            case 20: // greater
                this.push(Number(this.pop()<this.pop()));
            break;
            // 単項ビット演算
            case 21: // notb
                this.push(~this.pop());
            break;
            // 単項論理演算
            case 22: // not
                this.push(Number(!this.pop()));
            break;

            // 入出力命令
            case 23: // out スタックトップの値を出力
                console.log("[output],",this.pop().toString(16));
            break;
            default:
            break;
        }
        this.#progcnt++;
        return this;
    }
    runall() { // 最後まで命令を実行する(最大100000)
        let cnt = 0;
        while (cnt<100000&&!this.endRunning()) {cnt++;this.next();}
        return this;
    }
    check(file) {
        if (!(new TextDecoder("utf-8")).decode(new Uint8Array(file)).startsWith("nveof\n")) {
            throw "This file is not NVE-OF";
        }
    }
    endRunning() {return this.#progcnt>=this.size}
}

let path = process.argv[2];
if (path==null) {
    console.log("Failed to open file.");
    return;
}
fs.stat( path , ( er , stat ) => {
    if( er ){
        console.log("Failed to open file.");
        return;
    }else{
        if (stat.isFile()) {
            console.log("File:",path);
            let a = new NVE(fs.readFileSync(path));
            a.runall();
            // console.log(a.getData())
        }
        else {
            console.log("Failed to open file.")
            return;
        }
    }
});